File: packages/mesh-io/typescript/src/read-point-set.ts
import {
  BinaryFile,
  PointSet,
  getFileExtension,
  WorkerPoolFunctionResult,
  WorkerPoolFunctionOption,
} from "itk-wasm";

import mimeToPointSetIo from "./mime-to-point-set-io.js";
import extensionToPointSetIo from "./extension-to-point-set-io.js";
import pointSetIoIndex from "./point-set-io-index.js";

import ReadPointSetOptions from "./read-point-set-options.js";
import ReadPointSetResult from "./read-point-set-result.js";

interface ReaderResult extends WorkerPoolFunctionResult {
  couldRead: boolean;
  pointSet: PointSet;
}
interface ReaderOptions extends WorkerPoolFunctionOption {
  /** Only read point set metadata -- do not read pixel data. */
  informationOnly?: boolean;
}
type Reader = (
  serializedPointSet: File | BinaryFile,
  options: ReaderOptions,
) => Promise<ReaderResult>;

/**
 * Read a point set file format and convert it to the itk-wasm file format
 *
 * @param {webWorker} null | webWorker - Web worker to run the pipeline or null to run it in a new worker
 * @param {File | BinaryFile} serializedPointSet - Input point set serialized in the file format
 * @param {ReadPointSetOptions} options - options to cast the resulting point set type or to only read point set metadata
 *
 * @returns {Promise<ReadPointSetResult>} - result object with the point set and the web worker used
 */
async function readPointSet(
  serializedPointSet: File | BinaryFile,
  options: ReadPointSetOptions = {},
): Promise<ReadPointSetResult> {
  const mimeType = (serializedPointSet as File).type ?? "";
  const fileName =
    (serializedPointSet as File).name ??
    (serializedPointSet as BinaryFile).path ??
    "fileName";
  const extension = getFileExtension(fileName).toLowerCase();
  let usedWebWorker = options?.webWorker;

  let serializedPointSetFile = serializedPointSet as BinaryFile;
  if (serializedPointSet instanceof Blob) {
    const serializedPointSetBuffer = await serializedPointSet.arrayBuffer();
    serializedPointSetFile = {
      path: serializedPointSet.name,
      data: new Uint8Array(serializedPointSetBuffer),
    };
  }

  let io = null;
  if (mimeType && mimeToPointSetIo.has(mimeType)) {
    io = mimeToPointSetIo.get(mimeType);
  } else if (extensionToPointSetIo.has(extension)) {
    io = extensionToPointSetIo.get(extension);
  } else {
    for (const readerWriter of pointSetIoIndex.values()) {
      if (readerWriter[0] !== null) {
        let {
          webWorker: testWebWorker,
          couldRead,
          pointSet,
        } = await (readerWriter[0] as unknown as Reader)(
          {
            path: serializedPointSetFile.path,
            data: serializedPointSetFile.data.slice(),
          },
          {
            informationOnly: options.informationOnly,
            webWorker: usedWebWorker,
            noCopy: options?.noCopy,
          },
        );
        usedWebWorker = testWebWorker;
        if (couldRead) {
          return { webWorker: usedWebWorker, pointSet };
        }
      }
    }
  }
  if (!io) {
    throw Error("Could not find IO for: " + fileName);
  }
  const readerWriter = pointSetIoIndex.get(io as string);

  const reader = (readerWriter as Array<Reader>)[0];
  let {
    webWorker: testWebWorker,
    couldRead,
    pointSet,
  } = await reader(serializedPointSetFile, {
    informationOnly: options.informationOnly,
    webWorker: usedWebWorker,
    noCopy: options?.noCopy,
  });
  usedWebWorker = testWebWorker;
  if (!couldRead) {
    throw Error("Could not read: " + fileName);
  }

  return { webWorker: usedWebWorker, pointSet };
}

export default readPointSet;

File: packages/mesh-io/python/itkwasm-mesh-io-wasi/itkwasm_mesh_io_wasi/read_point_set.py
import os
import importlib
from pathlib import Path

from itkwasm import PointSet

from .extension_to_point_set_io import extension_to_point_set_io
from .point_set_io_index import point_set_io_index

def read_point_set(
    serialized_point_set: os.PathLike,
    information_only: bool = False,
) -> PointSet:
    """Read a point set file format and convert it to the ITK-Wasm file format.

    :param serialized_point_set: Input point set serialized in the file format
    :type  serialized_point_set: os.PathLike

    :param information_only: Only read point set metadata -- do not read pixel data.
    :type  information_only: bool

    :return: Output point set
    :rtype:  PointSet
    """
    extension = ''.join(Path(serialized_point_set).suffixes)

    io = None
    if extension in extension_to_point_set_io:
        func = f"{extension_to_point_set_io[extension]}_read_point_set"
        mod_name = f"itkwasm_mesh_io_wasi.{func}"
        mod = importlib.import_module(mod_name)
        io = getattr(mod, func)
    else:
        for ioname in point_set_io_index:
            func = f"{ioname}_read_point_set"
            mod_name = f"itkwasm_mesh_io_wasi.{func}"
            mod = importlib.import_module(mod_name)
            io = getattr(mod, func)
            could_read, point_set = io(serialized_point_set, information_only=information_only)
            if could_read:
                return point_set

    if io is None:
        raise RuntimeError(f"Could not find an point set reader for {extension}")

    could_read, point_set = io(serialized_point_set, information_only=information_only)
    if not could_read:
        raise RuntimeError(f"Could not read {serialized_point_set}")

    return point_set


def pointsetread(
    serialized_point_set: os.PathLike,
    information_only: bool = False,
) -> PointSet:
    return read_point_set(serialized_point_set, information_only=information_only)

pointsetread.__doc__ = f"""{read_point_set.__doc__}
    Alias for read_point_set.
    """